### PR TITLE
MDEV-29955: Set path for zlib library with pkg-config

### PR DESCRIPTION
**Pull request by keeper**
- [x] *The Jira issue number for this PR is: MDEV-29955*

## Description
`FindZLIB` module uses variable `ZLIB_ROOT`[1] to look for libraries. By
setting the variable, `FindZLIB` is able to search the libraries that
installed in a non-system path (/workspace/mylib for example).

And when using `z` in `LINK_LIBRARIES()` CMake tries to lookup the
library in system path by default. It doesn't work if the library isn't
installed in the path, and use ${ZLIB_LIBRARY} which set by FindZLIB
solve the issue.

[1]: https://cmake.org/cmake/help/latest/module/FindZLIB.html#hints

## How can this PR be tested?
* [x] Builds with no errors with Zlib installed in custom path

Manual checks:
1. pkg-config:
```
$ pkg-config --variable=libdir zlib
/workspace/mylib/lib
```

2. Generated CMake build file
```
$ cmake -S . -B build -G Ninja --fresh -DWIHT_ZLIB=system
...
...
$ grep libz.so build/build.ninja
...
...
LINK_LIBRARIES = -Wl,-rpath,/workspace/mylib/lib: ... /workspace/mylib/lib/libz.so  -lm  -ldl
```

If the changes are not amenable to automated testing, please explain why not and carefully describe how to test manually.

## Basing the PR against the correct MariaDB version
- [x] *This is a bug fix and the PR is based against the earliest maintained branch in which the bug can be reproduced.*

## PR quality check
- [x] I checked the [CODING_STANDARDS.md](https://github.com/MariaDB/server/blob/11.0/CODING_STANDARDS.md) file and my PR conforms to this where appropriate.
- [x] For any trivial modifications to the PR, I am ok with the reviewer making the changes themselves.

## Copyright
All new code of the whole pull request, including one or several files
that are either new files or modified ones, are contributed under the
BSD-new license. I am contributing on behalf of my employer Amazon Web
Services.

### CODE REVIEW

File: cmake/zlib.cmake [hunk 1]
```diff
@@ -37,6 +37,8 @@ MACRO (MYSQL_CHECK_ZLIB_WITH_COMPRESS)
   IF(WITH_ZLIB STREQUAL "bundled")
     MYSQL_USE_BUNDLED_ZLIB()
   ELSE()
+    FIND_PACKAGE(PkgConfig QUIET)
```
Comment (grooverdan) [inline cmake/zlib.cmake:40]: is, to depend on the finding of pkg-config is the following required:
```
IF(PKG_CONFIG_FOUND)
  PKG_GET_VARIABLE(ZLIB_ROOT zlib prefix)
ENDIF()
```

I'm roughly thinking it is based on https://gitlab.kitware.com/cmake/cmake/-/blob/master/Modules/FindPkgConfig.cmake#L131 .

Have you tried with pkg-config not there?

File: cmake/zlib.cmake [hunk 2]
```diff
@@ -37,6 +37,8 @@ MACRO (MYSQL_CHECK_ZLIB_WITH_COMPRESS)
...
+    PKG_GET_VARIABLE(ZLIB_ROOT zlib prefix)
     INCLUDE(FindZLIB)
```
Comment (grooverdan) [inline cmake/zlib.cmake:42]: odd, but not your fault, not sure why this isn't `find_package(zlib)`

File: storage/maria/CMakeLists.txt [hunk 3]
```diff
@@ -124,7 +124,7 @@ ENDIF()
 IF (CURL_FOUND)
   INCLUDE_DIRECTORIES(${CURL_INCLUDE_DIRS})
   MYSQL_ADD_PLUGIN(s3 ha_s3.cc ${S3_SOURCES} COMPONENT s3-engine
-    LINK_LIBRARIES ${CURL_LIBRARIES} z STORAGE_ENGINE NOT_EMBEDDED CONFIG s3.cnf)
```
Comment (grooverdan) [inline storage/maria/CMakeLists.txt:127]: good. Thank you.

File: cmake/zlib.cmake [hunk 4]
```diff
@@ -37,6 +37,8 @@ MACRO (MYSQL_CHECK_ZLIB_WITH_COMPRESS)
...
     IF(ZLIB_FOUND)
      INCLUDE(CheckFunctionExists)
```
Comment (grooverdan) [inline cmake/zlib.cmake:44]: Is  `SET(CMAKE_REQUIRED_LIBRARIES ${ZLIB_LIBRARIES})`  required below?

File: cmake/zlib.cmake (hunk 1 above)
Comment (keeper) [reply cmake/zlib.cmake:40]: > Have you tried with pkg-config not there?
Interesting, I moved `pkg-config` from `/usr/bin` to somewhere not in $PATH but the command can still find the right path:

```
-- Found ZLIB: /workspace/mylib/lib/libz.so (found version "1.2.13") 
```

I basically follow the convention here:
https://github.com/MariaDB/server/commit/3a33ae86010051a1022c392f2936b5fc9f3ab19e#diff-8e35291051929131abcdfd960c0e4e246317876c5ec2989154a8e810d489c7d6R81-R82, do you prefer me checking the availability here?

File: cmake/zlib.cmake (hunk 2 above)
Comment (keeper) [reply cmake/zlib.cmake:42]: Thank you for pointing out, I can also change the line to use `find_package()` which is more conscience

File: cmake/zlib.cmake (hunk 4 above)
Comment (keeper) [reply cmake/zlib.cmake:44]: Yes - thanks again for pointing this out

File: cmake/zlib.cmake (hunk 1 above)
Comment (grooverdan) [reply cmake/zlib.cmake:40]: I think its fine not that you've tested it. Thanks for the pcre reference implementation.

**Review by grooverdan — APPROVED** [review]
`find_package(zlib)`

> Thank you for pointing out, I can also change the line to use find_package() which is more conscience

Optional, if you have time, sure happy to take the improvement (and hope it passes all of CI).

**Review by vaintroub — CHANGES_REQUESTED** [review]

File: cmake/zlib.cmake (hunk 2 above)
Comment (vaintroub) [reply cmake/zlib.cmake]: What happens if zlib is not installed. What is the value of ZLIB_ROOT?

File: storage/maria/CMakeLists.txt [hunk 5]
```diff
@@ -124,7 +124,7 @@ ENDIF()
...
+    LINK_LIBRARIES ${CURL_LIBRARIES} ${ZLIB_LIBRARY} STORAGE_ENGINE NOT_EMBEDDED CONFIG s3.cnf)
```
[the remark this answers is not in the document]
Comment (keeper) [reply storage/maria/CMakeLists.txt:127]: > I would prefer checking availability and I would also prefer to exclude this on Windows
This has been fixed in latest update, thank you for the great advice.

> ZLIB_LIBRARY is not even documented, though appears to work
Totally agree. I also tends to use `ZLIB_LIBRARIES` but it's `ZLIB_LIBRARY` everywhere. To keep the consistency across codes I'll leave it as is per your suggestion.

**Review by vaintroub — APPROVED** [review]
Looks good to me, thanks!

**Pull request merged**